### ISSUE TITLE
Accessing a detached overlay view (android) with Appium

### ISSUE THREAD

**Issue by DvirSh** [appium/appium#7722]
## **TL;DR:**
It would be great if Appium would have a command to find and even access overlays (android views that were added dynamically with WindowManager).

## **Longer version:**
I have an android application in which I pop up an overlay window. This is done by calling:
```java
myWebView.loadUrl("someUrl");
windowManager = (WindowManager) context.getSystemService(WINDOW_SERVICE);
windowManager.addView(myWebView, layout);
```

And I want to test that the overlay is displayed successfully, and also that I can perform actions in it (example scenario: there is an X button inside that web view, and clicking it should close the overlay).

I tried to use Appium inspector / android uiautomator and neither was able to detect the webview on the screen at all.
I've also tried to use ```myWebView.setContentDescription("some_tag")``` to label the view and then access it with ```driver.findElement("//*[@content-desc='some_tag']")``` but it didn't find it.

I've found that the following ADB command can indicate wether the overlay is present:
```
adb shell dumpsys window windows | grep "package=myPackage appop=SYSTEM_ALERT_WINDOW"
```

But this would be a fragile solution, because I will probably be counting on running an external process in my test, not to mention that it depends on grep (can be easily transferred to JS).

I couldn't find anything similar in Appium.
Can this ability be added? To query for WindowManager's dynamically added views, and even provide selector-based access to them?

**Comment by jlipps** [comment]
@DvirSh does anything about this overlay window show up in the page source while it's present? Also, the preferred way to find an element by content-desc is to try and find the element by accessbilityId, not using an xpath query.

**Comment by DvirSh** [comment]
@jlipps What do you mean by "page source"? (I think it is quite ambiguous)

About the content-desc, you mean I should set the content-desc in my Java code (as above), but on my JS test I should use ```driver. findElementByAccessibilityId("some_tag")```? I'll try it out...

**Comment by DvirSh** [comment]
No, I tried using ```driver.click("~some_tag")``` and it could not find the element (although it's on the screen)

**Comment by jlipps** [comment]
@DvirSh every appium and selenium client has a command to retrieve the page source, which is an XML document. I don't know what library you're using so I can't comment on the specific command you should use, but it would be in the library docs.

> About the content-desc, you mean I should set the content-desc in my Java code (as above), but on my JS test I should use driver. findElementByAccessibilityId("some_tag")? I'll try it out...

Yes.

**Comment by WitoldKijKijowieckiMobica** [comment]
Did you try driver.switch_to.context(contexts[-1])

**Comment by jlipps** [comment]
closing for no response. but yes if a webview is present, switching contexts is the right solution.

**Comment by naveedabdullah** [comment]
If you are creating a new Instance of WindowManager, you need to make sure the Flag "FLAG_NOT_FOCUSABLE" is removed

public  WindowManager.LayoutParams handleParams;
handleParams = getLayoutParams();

handleParams.flags = handleParams.flags & ~WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE

private static WindowManager.LayoutParams getLayoutParams(Context context) {
            return new WindowManager.LayoutParams(
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    WindowManager.LayoutParams.TYPE_SYSTEM_ERROR,
                    WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE |
                            WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH |
                            WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL |
                            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS |
                            WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                            | WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD,
                    PixelFormat.TRANSLUCENT);
        }

This works for me 

but when i try to retrieve a Window from getWindow(), i am not able to click on the window (though i can detect it). I am trying adding flags and removing flags. 
Lets see if that helps

**Comment by fruwe** [comment]
@jlipps Hi, I have the same problem.

I have an overlay button, which is visible in all apps and belongs to let's call it app1. 
In a seperate test application (app2) the overlay button of app1 is always visible. I would like to press that button with appium.

The page_source method returns everything from the test app. However the overlay button is not included in the source. I tried this with UIAutomator and UIAutomator2.

Furthermore the available_contexts method return only one item in UIAutomator and two items in UIAutomator2, however switching the the the second context caused the test to freeze. In any case, I am not using webviews anyway, so I guess context switching is nothing I should be concerned with.

Searching the element also did not work:
```ruby
@driver.find_element(id: "com.xxx.yyy:id/zzzButton")
```
However, I can touch the button with a touch action like so:

```ruby
::Appium::TouchAction.new(@driver).tap(x:1900, y: 500).perform
```

I am using Appium v1.10.1 (Not "Appium Desktop" since it is not working with UIAutomator2!). My doctor's result is:

```
Appium Doctor v.1.7.0
### Diagnostic for necessary dependencies starting ###
 ✔ The Node.js binary was found at: /home/chris/.nvm/versions/node/v10.14.1/bin/node
 ✔ Node version is 10.14.1
 ✔ ANDROID_HOME is set to: /home/chris/Android/Sdk
 ✔ JAVA_HOME is set to: /usr/lib/jvm/java-8-openjdk-amd64
 ✔ adb exists at: /home/chris/Android/Sdk/platform-tools/adb
 ✔ android exists at: /home/chris/Android/Sdk/tools/android
 ✔ emulator exists at: /home/chris/Android/Sdk/tools/emulator
 ✔ Bin directory of $JAVA_HOME is set
### Diagnostic for necessary dependencies completed, no fix needed. ###

### Diagnostic for optional dependencies starting ###
 ✔ opencv4nodejs is installed.
### Diagnostic for optional dependencies completed, no fix needed. ###

Everything looks good, bye!
```

Is there any way to make that button clickable without putting coordinates in my code?

**Comment by simon30rus** [comment]
@fruwe any chance did you find the solution? I am facing exactly the same issue. I tried removing WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE from my view overlay that causes UIAutomator to see the overlay view, but the rest gets lost...
